tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,450 @@
+	tb.beaconsCh[epoch] <- beacon
+	close(tb.beaconsCh[epoch])
+	tb.beaconsMu.Unlock()
+}
+
+func (tb *TortoiseBeacon) handleMessage(m Message) error {
+	switch m := m.(type) {
+	case ProposalMessage:
+		return tb.handleProposalMessage(m)
+	case VotingMessage:
+		return tb.handleVotingMessage(m)
+	case BatchVotingMessage:
+		return tb.handleBatchVotingMessage(m)
+	default:
+		return ErrUnknownMessageType
+	}
+}
+
+func (tb *TortoiseBeacon) handleProposalMessage(m ProposalMessage) error {
+	epoch := m.Epoch()
+	hash := m.Hash()
+
+	mt := tb.classifyMessage(m, epoch)
+	switch mt {
+	case TimelyMessage:
+		tb.timelyProposalsMu.Lock()
+
+		tb.timelyProposalsMap[hash] = m.Proposals()
+		tb.timelyProposalsList[epoch] = append(tb.timelyProposalsList[epoch], m.Proposals())
+
+		tb.timelyProposalsMu.Unlock()
+
+		return nil
+	default:
+		// TODO(nkryuchkov): handle other types
+		return nil
+	}
+}
+
+func (tb *TortoiseBeacon) handleVotingMessage(m VotingMessage) error {
+	epoch := m.Epoch()
+
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	pair := EpochRoundPair{
+		EpochID: epoch,
+		Round:   currentRound,
+	}
+
+	mt := tb.classifyMessage(m, epoch)
+	switch mt {
+	case TimelyMessage:
+		tb.votesMu.Lock()
+		tb.votes[pair] = append(tb.votes[pair], m.Hash())
+		tb.votesMu.Unlock()
+
+		return tb.beaconCalculator.CountVotes(m) // TODO(nkryuchkov): count votes properly
+
+	default:
+		// TODO(nkryuchkov): handle other types
+		return nil
+	}
+}
+
+func (tb *TortoiseBeacon) handleBatchVotingMessage(m BatchVotingMessage) error {
+	epoch := m.Epoch()
+
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	pair := EpochRoundPair{
+		EpochID: epoch,
+		Round:   currentRound,
+	}
+
+	mt := tb.classifyMessage(m, epoch)
+	switch mt {
+	case TimelyMessage:
+		for _, hash := range m.HashList() {
+			tb.votesMu.Lock()
+			tb.votes[pair] = append(tb.votes[pair], hash)
+			tb.votesMu.Unlock()
+		}
+
+		return tb.beaconCalculator.CountBatchVotes(m) // TODO(nkryuchkov): count votes properly
+
+	default:
+		// TODO(nkryuchkov): handle other types
+		return nil
+	}
+}
+
+func (tb *TortoiseBeacon) classifyMessage(m Message, epoch types.EpochID) MessageType {
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	round := 0
+	if m, ok := m.(VotingMessage); ok {
+		round = m.Round()
+	}
+
+	switch {
+	case round <= currentRound-1:
+		return TimelyMessage
+	case round == currentRound-2:
+		return DelayedMessage
+	default:
+		return LateMessage
+	}
+}
+
+// for K rounds : in each round that lasts δ, wait for proposals to come in
+func (tb *TortoiseBeacon) roundTicker(epoch types.EpochID) {
+	ticker := time.NewTicker(tb.networkDelta)
+	defer ticker.Stop()
+
+	tb.currentRoundsMu.Lock()
+	tb.currentRounds[epoch] = 0
+	tb.currentRoundsMu.Unlock()
+
+	// round 0 already happened at this point, starting from round 1
```

Comment:
what should happen in round 0?

Reply:
@antonlerner 
1. take all ATXs received in last epoch (i -1)
2. concat them into a single proposal message and sign it
3. send the proposal message